packages/concerto-core/lib/serializer/jsongenerator.js

```diff
@@ -142,18 +144,40 @@ class JSONGenerator {
             }
         }
 
+        // by default we set the $class on the result
         result.$class = classDeclaration.getFullyQualifiedName();
+
+        // if we are in the inferClass mode and this is not the root instance
+        // we attempt to remove the $class
+        if(this.inferClass && !parameters.isRoot && parameters.field) {
+            const fieldClassDecl = parameters.modelManager.getType(parameters.field.getFullyQualifiedTypeName());
+            // if the $class cannot be unambigiously inferred from the type of the field in the model
+            // we have to include it, but we attempt to shorten it, if the object and the field are in the same ns
+            if(fieldClassDecl.getAssignableClassDeclarations().length > 2) {
+                const objAndFieldSameNs = parameters.field ? ModelUtil.getNamespace(parameters.field.getFullyQualifiedTypeName()) === obj.getNamespace() : false;
+                result.$class = (objAndFieldSameNs && !parameters.isRoot)  ? obj.getType() : obj.getFullyQualifiedType();
+            }
+            else {
+                // we don't need the $class - we can recreate it from the model
+                delete result.$class;
+            }
+        }
+
         if(this.deduplicateResources && id) {
             result.$id = id;
         }
 
+        // no longer dealing with the root object...
+        parameters.isRoot = false;
+
         // Walk each property of the class declaration
         const properties = classDeclaration.getProperties();
         for (let index in properties) {
             const property = properties[index];
             const value = obj[property.getName()];
             if (!Util.isNull(value)) {
                 parameters.stack.push(value);
+                parameters.field = property; // save the property to the stack!
```

Comment:
Where do we pop this value?

Reply:
The comment is misleading, this is not pushed to the stack, it is just saved to parameters.